### PR TITLE
Full package restructure

### PR DESCRIPTION
**Pull request by steven-murray**
Full package restructure.

Used `pyscaffold` to set up the structure. Adds

* pre-commit
* black
* flake8
* tox
* pytest / tests skeleton
* setuptools-scm
* docs skeleton
* python 3 compatibility

### CODE REVIEW

File: .isort.cfg [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[settings]
```
Comment (plaplant) [inline .isort.cfg:1]: What does this config file control?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    args: ['--fix=no']
+  - id: flake8
+    args: ['--max-line-length=88']  # default of Black
+
+- repo: https://github.com/pre-commit/mirrors-isort
+  rev: v4.3.4
+  hooks:
+  - id: isort
+
+-   repo: https://github.com/ambv/black
```
Comment (plaplant) [inline .pre-commit-config.yaml]: It seems like this is indented more than the previous hooks. Is this on purpose?

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# THIS SCRIPT IS SUPPOSED TO BE AN EXAMPLE. MODIFY IT ACCORDING TO YOUR NEEDS!
+
+sudo: false
+language: python
+virtualenv:
+  system_site_packages: false
+matrix:
+  fast_finish: true
+  include:
+    - python: 3.6
```
Comment (plaplant) [inline .travis.yml:12]: I think we also want to test building and running on python3.7, but I can set that up in a different PR once this is merged in and we have proper tests to run.

File: AUTHORS.rst [hunk 4]
```diff
@@ -0,0 +1,5 @@
+============
+Contributors
+============
+
+* Steven Murray <steven.g.murray@asu.edu>
```
Comment (plaplant) [inline AUTHORS.rst:6]: We should probably add Zac to this list too 😉

File: CHANGELOG.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
+=========
```
Comment (plaplant) [inline CHANGELOG.rst:1]: Changelog from Day 1! Love it 😄

File: LICENSE.txt [hunk 6]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2019 Steven Murray
```
Comment (plaplant) [inline LICENSE.txt:3]: I'm not super familiar with how software copyright works. Should this be copyrighted by "the authors"? Or our group? I know things like pyuvdata are copyright "The Radio Astronomy Software Group", but not sure how that plays out in practice. (I'm also not sure if I'm making a big deal out of nothing, so feel free to ignore this...)

File: README.rst [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+`FFTW_PATH=/usr/lib pip install .`.
+
+Other dependencies of spin1_beam_model are installed automatically. However, if you are
+using `conda` and would like these packages to be installed with `conda` rather than
+`pip`, you may want to do `conda install numpy h5py scipy numba cffi` prior to
+installing the package.
+
+Quick Usage
+===========
+
```
Comment (plaplant) [inline README.rst:32]: We probably want to put something here, but we can do that as part of another PR.

File: docs/index.rst [hunk 8]
```diff
@@ -0,0 +1,59 @@
+================
+spin1_beam_model
+================
+
+This is the documentation of **spin1_beam_model**.
+
+.. note::
+
+    This is the main page of your project's `Sphinx`_ documentation.
```
Comment (plaplant) [inline docs/index.rst:9]: Should we edit this text at all? I'm not familiar with doing formal documentation like this, so forgive the naive question...

File: requirements.txt [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# =============================================================================
```
Comment (plaplant) [inline requirements.txt:1]: If we're not actually using the `requirements.txt`, should we still have it in the repo?

File: setup.cfg [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+author = Zachary Martinot
+author-email = zmarti@sas.upenn.edu
+license = mit
+long-description = file: README.rst
+long-description-content-type = text/x-rst; charset=UTF-8
+url = https://github.com/UPennEoR/spin1_beam_model
+project-urls =
+    Documentation = https://pyscaffold.org/
+# Change if running only on Windows, Mac or Linux (comma-separated)
+platforms = any
```
Comment (plaplant) [inline setup.cfg]: I think the ssht dependency means that we're only compativle with Mac and Linux...

File: src/spin1_beam_model/cst_processing.py [hunk 11]
```diff
@@ -73,43 +85,59 @@ def __init__(self, data_files, data_freqs, spatial_sampling, data_uncertainty,
 
         data_file = np.loadtxt(data_files[0], skiprows=2)
 
-        self.Ntheta = int(181./spatial_sampling)
-        self.Nphi = int(360./spatial_sampling)
+        self.Ntheta = int(181.0 / spatial_sampling)
```
Comment (plaplant) [inline src/spin1_beam_model/cst_processing.py:88]: Probably not the right place to ask this (and @zacharymartinot would be the one to know), but why is this 181 instead of 180? Some kind of periodicity issue?

File: src/spin1_beam_model/plotting.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
+import cmocean
```
Comment (plaplant) [inline src/spin1_beam_model/plotting.py:1]: I don't think `cmocean` is listed anywhere as a formal dependency. We should either make this import optional or list it in the `setup.cfg`.

File: src/spin1_beam_model/plotting.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+    return p
+
+
+def MuellerMatrixElement(J, i, j):
+
+    Pi = StokesMatrix(i)
+    Pj = StokesMatrix(j)
+
+    M_ij = np.einsum("...ab,...bc,...cd,...ad", Pi, J, Pj, J.conj()) / 2.0
```
Comment (plaplant) [inline src/spin1_beam_model/plotting.py:30]: `einsum` notation is always frustrating and opaque to me, but adding explanatory comments should probably be part of a larger documentation push for the repo. I'll make an issue for better docs.

File: tests/test_skeleton.py [hunk 14]
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+
+import pytest
+
+from spin1_beam_model.skeleton import fib
```
Comment (plaplant) [inline tests/test_skeleton.py]: Where is this function defined?

**Review by plaplant — COMMENTED** [review]
Thanks for setting up the infrastructure @steven-murray! Most of my comments are small, and are probably best dealt with by making issues and addressing them in separate PRs.

File: .isort.cfg (hunk 1 above)
Comment (steven-murray) [reply .isort.cfg:1]: The isort tool -- it was setup automatically by `pyscaffold`. I think it is kind of like black for your imports.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (steven-murray) [reply .pre-commit-config.yaml]: Nope. I'll fix that. Doesn't seem to break things though.

File: .travis.yml (hunk 3 above)
Comment (steven-murray) [reply .travis.yml:12]: Yeah... the whole .travis.yml is complete trash -- I figure you'll fix that ;-)

File: AUTHORS.rst (hunk 4 above)
Comment (steven-murray) [reply AUTHORS.rst:6]: Yup, done.

File: setup.cfg (hunk 10 above)
Comment (steven-murray) [reply setup.cfg]: Good catch. Fixed that.

File: src/spin1_beam_model/plotting.py (hunk 12 above)
Comment (steven-murray) [reply src/spin1_beam_model/plotting.py:1]: It's in `extras_require`, in `plots`. I thought it seemed rather like an appendage, not core, so I thought it better there. What do you think?

File: src/spin1_beam_model/plotting.py (hunk 13 above)
Comment (steven-murray) [reply src/spin1_beam_model/plotting.py:30]: Yes, let's definitely make that issue.

File: tests/test_skeleton.py (hunk 14 above)
Comment (steven-murray) [reply tests/test_skeleton.py]: Ugh. Sorry, this file shouldn't exist.

File: .travis.yml (hunk 3 above)
Comment (plaplant) [reply .travis.yml:12]: Fair enough :)

File: src/spin1_beam_model/plotting.py (hunk 12 above)
Comment (plaplant) [reply src/spin1_beam_model/plotting.py:1]: Ah okay. I didn't realize it's in the `setup.cfg` file. I think that's fine then.

**Review by plaplant — APPROVED** [review]
This looks awesome, thanks @steven-murray. I'm merging this in so we can start tackling some of the downstream issues in RIMEz.
